File: Eclipse/src/trabalhoSO/AlgoritmoCircular.java
package trabalhoSO;
import java.util.Scanner;
import javax.swing.JOptionPane;

public class AlgoritmoCircular {
	
	public static void main(String[] args) {
		
		Scanner input = new Scanner(System.in);
		
		int quantum, trocaContexto;
		String[] listaProgramas;
		String[] vetorString;
		
		String str = lerString("Digite os tempos de processador dos programas separando por espaço:");
		listaProgramas = str.split(" ");
		str = lerString("Digite quantum e fator de troca de contexto separados por espaço:\nQuantidade de programas a serem computados: " + listaProgramas.length);
		vetorString = str.split(" ");
		quantum = Integer.parseInt(vetorString[0]);
		trocaContexto = Integer.parseInt(vetorString[1]);
		int turnaround = tempoMedioTurnaround(listaProgramas, quantum, trocaContexto);
		msg("Tempo medio de turnaround: " + turnaround);
	}
	
	static int tempoMedioTurnaround(String[] lista, int quantum, int trocaContexto){
		
		int i;
		int tempo = 0;
		int[] listaProgramas = new int[lista.length];
		
		for (i = 0; i < lista.length; i++) {
			listaProgramas[i] = Integer.parseInt(lista[i]);
		}
		
		while(true){
			
			int count = 0;
			for (int j = 0; j < listaProgramas.length; j++)
				{if (listaProgramas[j] == 0){count++;}}
			System.out.println("Count: " + count);
			if (count == listaProgramas.length){break;}
			
			for (i = 0; i < listaProgramas.length; i++) {
				
				System.out.println("P" + i + " // Tempo: " + tempo + " // Valor: " + listaProgramas[i]);
				
				if(trocaContexto == 0){trocaContexto = 1;}
				
				if(listaProgramas[i] > 0){
					tempo += quantum + trocaContexto;
					if((listaProgramas[i] - quantum) < 0){
						listaProgramas[i] = Math.abs(listaProgramas[i] - quantum);
						quantum = listaProgramas[i];
					}else{
						listaProgramas[i] = listaProgramas[i] - quantum;
					}
				}
			}
		}
		
		System.out.println("Tempo = " + tempo + " // Tempo - Troca de Contexto = " + (tempo - trocaContexto));
		tempo = ((tempo - trocaContexto) / listaProgramas.length);
		
		return tempo;
	}
	
	static void msg(String str) {
		JOptionPane.showMessageDialog(null, str);
	}

	static int lerInteiro(String str) {
		return Integer.parseInt(JOptionPane.showInputDialog(str));
	}
	
	static String lerString(String str) {
		return JOptionPane.showInputDialog(str);
	}
}

File: Python/algoritmoCircular.py
# -*- coding: cp1252 -*-

def _cmp_by_program(x, y):
    return cmp(x[0], y[0])

def calcularTempo(lista, quantum, trocaContexto):
    origQuantum = quantum
    tempo = 0
    loopCount = 0
    vetorTempos = [0]
    vetorProgramas = []
    while(True):
        
        count = 0;
        for j in range(len(lista)):
            if (lista[j] == 0):
                    count += 1
        if (count == len(lista)):
            break

        i = 0
        while(i < len(lista)):
            
            #print "Começo -- Processo:", i+1, "// Tempo:", tempo, "// Valor:", lista[i]

            if (lista[i] > 0):
                if ((lista[i] - quantum) < 0):

                    oldTempo = tempo
                    tempo += lista[i] + trocaContexto
                    vetorTempos += [tempo]
                    lista[i] -= lista[i]
                    if (lista[i] == 0):
                        vetorProgramas += [[i+1, tempo]]  
                else:

                    oldTempo = tempo
                    if (loopCount == 0 and i == 0):
                        tempo += quantum
                        vetorTempos += [tempo]
                        lista[i] -= quantum
                    else:
                        tempo += quantum + trocaContexto
                        vetorTempos += [tempo]
                        lista[i] -= quantum
                    if (lista[i] == 0):
                        vetorProgramas += [[i+1, tempo]]

                print "Tempo:", tempo, "Processo:", i, "Valor:", lista[i]

            #print "Final -- Processo:", i+1, "// Tempo:", tempo, "// Valor:", lista[i]
            i += 1

        loopCount += 1

    vetorProgramas.sort(_cmp_by_program)
    print "Tempo final:", vetorProgramas[len(vetorProgramas)-1][1]
    print "Vetor de tempos:", vetorTempos
    print "Vetor tempos finais de programas:", vetorProgramas

    tempo = 0
    for j in range(len(vetorProgramas)):
        tempo += float(vetorProgramas[j][1] - vetorTempos[j])
    tempo = tempo / len(vetorProgramas)
    print "Tempo medio de espera:", tempo

    
    tempo = 0
    for j in range(len(vetorProgramas)):
        tempo += float(vetorProgramas[j][1])
    tempo = tempo / len(vetorProgramas)
    print "Tempo medio de turnaround:", tempo

   
     
    
#listaProgramas = [int(i) for i in raw_input("Digite os tempos de processador dos programas separando por espaço: ").split(" ")]
#quantum, trocaContexto = [int(i) for i in raw_input("Digite quantum e fator de troca de contexto separados por espaço: ").split(" ")]
listaProgramas = [12, 16, 7, 9]
quantum = 6
trocaContexto = 3
calcularTempo(listaProgramas, quantum, trocaContexto)

